### PR TITLE
Guide contributions to ytt standard lib

### PR DESCRIPTION
**Pull request by pivotaljohn**

### CODE REVIEW

File: docs/dev.md [hunk 1]
```diff
@@ -62,3 +62,38 @@ Ultimately `get-ytt.io` runs on AWS Lambda, hence is wrapped with Lambda specifi
...
+
+When contributing to this library note:
+
+- **Begin with the API**
+  - before delving too deep into implementation, solidify the Starlark API (usually a conversation with a maintainer).
+  - this is codified in a couple of acceptance tests (`pkg/yamltemplate/filetests/ytt-library/...`)
+  - the goal is to provide a consistent, goal-aligned feature set.
+- **Design near Python and Go**
+  - Starlark is a dialect of Python: to maintain the value of that heritage, be inspired by Python
+  - much of the functionality that can wants to be exposed can come from the Go ecosystem: to keep solutions simple, be inspired by Go.
```
Comment (cari-lynn) [inline docs/dev.md]:
```suggestion
  - much of the functionality exposed can come from the Go ecosystem: to keep solutions simple, be inspired by Go.
```

File: docs/dev.md [hunk 2]
```diff
@@ -62,3 +62,38 @@ Ultimately `get-ytt.io` runs on AWS Lambda, hence is wrapped with Lambda specifi
...
+  - that said, this library is typically operating in a specific context: a hermetic and deterministic calculation of (YAML-formatted) configuration. Bend to adapt to that context.
+- **Offer string() rather than assume string encoding**
+  - concrete types whose instances template authors will handle should implement `UnconvertableStarlarkValue` rather than assume to convert to a string. 
+  - in the conversion hint, suggest they use `string()`.
+  - this preserves the ability to choose _other_ encoding strategies, later, if desired.
+- **Names**
+  - overall: look for conventions present in existing modules (e.g. `is_..()` for predicates,  `string()` to encode to string, ...)
+  - **module** — consider in combination with function names: be as concrete as possible, reveal intention, and avoid redundancy (e.g. `ip.parse_addr()` > `io.parse_ip_addr()`)
```
Comment (cppforlife) [inline docs/dev.md]: io -> ip

File: docs/dev.md [hunk 3]
```diff
@@ -62,3 +62,38 @@ Ultimately `get-ytt.io` runs on AWS Lambda, hence is wrapped with Lambda specifi
...
+  - concrete types whose instances template authors will handle should implement `UnconvertableStarlarkValue` rather than assume to convert to a string. 
+  - in the conversion hint, suggest they use `string()`.
+  - this preserves the ability to choose _other_ encoding strategies, later, if desired.
+- **Names**
+  - overall: look for conventions present in existing modules (e.g. `is_..()` for predicates,  `string()` to encode to string, ...)
+  - **module** — consider in combination with function names: be as concrete as possible, reveal intention, and avoid redundancy (e.g. `ip.parse_addr()` > `io.parse_ip_addr()`) 
+  - **functions** — typically verb and verb phrases work best.
+    - **built-in** — use the format "${module}.[${type_name}.]${function_name}" (this value is not yet used)
+    - Go definition of the Starlark function should almost always be the snakeCase format of the same name.
+  - **Type()** — use the format "@ytt:${module}.${type}"
```
Comment (cppforlife) [inline docs/dev.md]: add e.g. ...

File: docs/dev.md [hunk 4]
```diff
@@ -62,3 +62,38 @@ Ultimately `get-ytt.io` runs on AWS Lambda, hence is wrapped with Lambda specifi
...
+  - that said, this library is typically operating in a specific context: a hermetic and deterministic calculation of (YAML-formatted) configuration. Bend to adapt to that context.
+- **Offer string() rather than assume string encoding**
+  - concrete types whose instances template authors will handle should implement `UnconvertableStarlarkValue` rather than assume to convert to a string. 
+  - in the conversion hint, suggest they use `string()`.
+  - this preserves the ability to choose _other_ encoding strategies, later, if desired.
+- **Names**
+  - overall: look for conventions present in existing modules (e.g. `is_..()` for predicates,  `string()` to encode to string, ...)
+  - **module** — consider in combination with function names: be as concrete as possible, reveal intention, and avoid redundancy (e.g. `ip.parse_addr()` > `io.parse_ip_addr()`) 
+  - **functions** — typically verb and verb phrases work best.
+    - **built-in** — use the format "${module}.[${type_name}.]${function_name}" (this value is not yet used)
```
Comment (cppforlife) [inline docs/dev.md]: add e.g. ...

File: docs/dev.md (hunk 1 above)
Comment (pivotaljohn) [reply docs/dev.md]: Yup... clearly superfluous; thank you.

**Pull request merged**